### ISSUE TITLE
Pickling Error: can't pickle <class 'PyCapsule'>: it's not found as builtins.PyCapsule 

### ISSUE THREAD

**Issue by afylot** [uqfoundation/dill#93]
I have this problem with pickling data with dill. The problem only occurs on python 3.4. My code works with python 3.3 and python 2.7.

**Comment by matsjoyce** [comment]
Could you add some code to reproduce this, as it looks like PyCapsule shouldn't really leak out of any APIs?

**Comment by afylot** [comment]
I have written for you this short code that reproduces my error (it has the same data structure that I use in my code). This code works in python3.3.6 but not in python3.4.1

This same code (as the original one) also randomly throws other errors, instead of "can't pickle <PyCapsule>":

PicklingError: Can't pickle <class 'numpy.flagsobj'>: it's not found as numpy.flagsobj
or
PicklingError: Can't pickle <class 'numpy.int64'>: it's not the same object as numpy.int64

CODE:
---------------------------------------------------
import dill as pickle
from numpy.random import rand

class Spectra():
    def __init__(self, ref, ntd, mntd, npv, I100K, z):
        self.ref   = ref
        self.ntd   = ntd
        self.mntd  = mntd
        self.npv   = npv
        self.I100K = I100K
        self.z     = z

def main():
    
    ref         = load_compute()
    ntd         = load_compute()
    I100K       = load_compute()
    npv         = load_compute()
    z           = rand()
    mntd        = load_compute()

    spectra      = Spectra(ref, ntd, mntd, npv, I100K, z) 

    savefile = r'./save.pkl'
    savef = open(savefile,'wb')
    pickle.dump(spectra,savef)
    savef.close()
    return


    
def load_compute():
    spec     = short()

    spec['perc']        = compute()
    spec['perc5']       = compute()
    spec['perc50']      = compute()
    spec['perc95']      = compute()
    spec['percint']     = compute()
    spec['percintcorr'] = compute()
    spec['percnhi']     = compute()
    return spec

def short():
    cosmo, z = read_header_short()
    los      = read_all_spectra()
    cl       = read_lines()
    cl['z'] = z
    cl['gammahi'] = rand()
    return {'cosmo':cosmo, 'z':z, 'los':los, 'cl':cl}

def read_header_short():
    zqso = rand()
    
    cosmo = {'omegam':rand(), 'omegab':rand(),'omegal':rand(), 'h':rand(), 
                'ymass':rand(), 'mu':rand()}
    cosmo['omegam']  = rand()
    cosmo['omegab']  = rand()
    cosmo['omegal']  = rand()
    cosmo['h']       = rand()
    cosmo['mu']      = rand()
        
    return (cosmo, zqso)

def read_all_spectra():
    
    los={'T_real':rand(486699),
         'v':rand(486699),
         'wl':rand(486699),
         'od_real':rand(486699),
         'T_redshift':rand(486699),
         'tau':rand(486699),
         'F':rand(486699),
         'od_redshift':rand(486699)}
    return los

def read_lines():
    n=7131
    cl = {'start':rand(n), 
          'stop':rand(n), 
          'top':rand(n)}
    cl['v0']   = rand(n)
    cl['tau0'] = rand(n)
    cl['sig2'] = rand(n)
    cl['nhi_curv'] = rand(n)
    cl['nhi_int']  = rand(n)
    cl['nhi_int_correct']  = rand(n)
    cl['T_real']           = rand(n)
    cl['od_real']          = rand(n)
    cl['T_real_max']       = rand(n)
    cl['od_real_max']      = rand(n)
    cl['T_real_mean']      = rand(n)
    cl['od_real_mean']     = rand(n)
    cl['T_redshift']       = rand(n)
    cl['od_redshift']      = rand(n)
    cl['T_redshift_max']   = rand(n)
    cl['od_redshift_max']  = rand(n)
    cl['T_redshift_mean']  = rand(n)
    cl['od_redshift_mean'] = rand(n)
    cl['nhi_int']          = rand(n)
    cl['nhi_int_correct']  = rand(n)
    cl['T_real_max']       = rand(n)
    cl['od_real_max']      = rand(n)
    cl['T_real_mean']      = rand(n)
    cl['od_real_mean']     = rand(n)
    cl['T_redshift_max']   = rand(n)
    cl['od_redshift_max']  = rand(n)
    cl['T_redshift_mean']  = rand(n)
    cl['od_redshift_mean'] = rand(n)
    return cl

def compute():
    n = 19
    percs = {'xbin':rand(n),
                'sig2':rand(n), 
                'sigmasig2':rand(n),
                'nhi':rand(n), 
                'sigmanhi':rand(n),
                'od':rand(n),
                'sigmaod':rand(n),
                'T':rand(n),
                'sigmaT':rand(n)}
    return percs

**Comment by matsjoyce** [comment]
Seems to work with 3.4.3. Which numpy version do you use? I'm on 1.9.2.

**Comment by afylot** [comment]
numpy v 1.9.2

**Comment by mmckerns** [comment]
This also works for me with the same versions as both of you.

What's your OS and compiler?

**Comment by mmckerns** [comment]
I just noticed that you are on python `3.4.1`, while @matsjoyce and I are on `3.4.3`.  Could that be the issue?  @afylot: I'll take no response as "I've upgraded to `3.4.3` and it's no longer an issue".  I'll reopen if that's not the case.

**Comment by itdxer** [comment]
Got the same issue in Python 3.4.3 and NumPy 1.9.0

```
>>> import dill
>>> import numpy as np
>>> np.__version__
'1.9.0'
>>>
>>> class A(object):
...  pass
...
>>> a = A()
>>> dill.dumps(a)
>>> # Got valid output
>>>
>>> class B(object):
...  c = np.array([0])
...
>>> b = B()
>>> dill.dumps(b)
_pickle.PicklingError: Can't pickle <class 'PyCapsule'>: it's not found as builtins.PyCapsule
```

**Comment by mmckerns** [comment]
@itdxer: Thanks. Your version if `numpy` is `1.9.0` while it seems to work with `1.9.2` for us.  Can you tell us what your OS and compiler is?  Mine is:

    Python 3.4.3 (default, Mar 10 2015, 14:50:30) 
    [GCC 4.2.1 Compatible Apple Clang 4.1 ((tags/Apple/clang-421.11.66))] on darwin
    Type "help", "copyright", "credits" or "license" for more information.
    >>> 

I have been assuming that @matsjoyce and I aren't seeing the issue due to @afylot having older versions of `python`. However, if your issue is an older version of `numpy`, that's a little different, but not terrible.  If it's a platform issue, that's another thing.

**Comment by itdxer** [comment]
OS X Yosemite 10.10.3
```
$ python
Python 3.4.3 (default, Apr 13 2015, 00:05:38)
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.49)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>>
```

**Comment by mmckerns** [comment]
Hmm.  I'm still not seeing it, but I do see that it's also an open issue for `IPython`.

**Comment by mmckerns** [comment]
FYI, @itdxer: the `PyCapsule` API replaced the legacy `PyCObject` API (a while ago)… so this is a fundamental low-level object.  The change to `PyCapsule` happened a while back… so I'm a bit surprised to be only seeing this issue in `python3.4` and not earlier or in earlier versions… however, `python3.4` uses very different pickling.

I'll have to do some digging, and also work to reproduce the error you are seeing in this issue.

**Comment by mmckerns** [comment]
So, with my configuration, I see this:

    >>> import dill
    >>> import numpy as np
    >>> 
    >>> class B(object):
    ...   c = np.array([0])
    ... 
    >>> b = B()
    >>> dill.detect.trace(True)
    >>> dill.dumps(b)
    T2: <class '__main__.B'>
    F2: <function _create_type at 0x106bf9c80>
    T1: <class 'type'>
    F2: <function _load_type at 0x106bf9bf8>
    T1: <class 'object'>
    D2: <dict object at 0x106a5ea08>
    B1: <built-in function _reconstruct>
    F2: <function _get_attr at 0x106bff620>
    M2: <module 'numpy.core.multiarray' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/numpy/core/multiarray.so'>
    F2: <function _import_module at 0x106bff730>
    T4: <class 'numpy.ndarray'>
    T4: <class 'numpy.dtype'>
    b'\x80\x03cdill.dill\n_create_type\nq\x00(cdill.dill\n_load_type\nq\x01X\x04\x00\x00\x00typeq\x02\x85q\x03Rq\x04X\x01\x00\x00\x00Bq\x05h\x01X\x06\x00\x00\x00objectq\x06\x85q\x07Rq\x08\x85q\t}q\n(X\r\x00\x00\x00__slotnames__q\x0b]q\x0cX\n\x00\x00\x00__module__q\rX\x08\x00\x00\x00__main__q\x0eX\x07\x00\x00\x00__doc__q\x0fNX\x01\x00\x00\x00cq\x10cdill.dill\n_get_attr\nq\x11cdill.dill\n_import_module\nq\x12X\x15\x00\x00\x00numpy.core.multiarrayq\x13\x85q\x14Rq\x15X\x0c\x00\x00\x00_reconstructq\x16\x86q\x17Rq\x18cnumpy\nndarray\nq\x19K\x00\x85q\x1aC\x01bq\x1b\x87q\x1cRq\x1d(K\x01K\x01\x85q\x1ecnumpy\ndtype\nq\x1fX\x02\x00\x00\x00i8q K\x00K\x01\x87q!Rq"(K\x03X\x01\x00\x00\x00<q#NNNJ\xff\xff\xff\xffJ\xff\xff\xff\xffK\x00tq$b\x89C\x08\x00\x00\x00\x00\x00\x00\x00\x00q%tq&butq\'Rq()\x81q).'

Can you post what you get with `dill.detect.trace(True)`?   You could also investigate with `dill.detect.bad*`...

**Comment by itdxer** [comment]
```
>>> import dill
>>> import numpy as np
>>>
>>> class B(object):
...  c = np.array([0])
...
>>> b = B()
>>> dill.detect.trace(True)
>>> dill.dumps(b)
T2: <class '__main__.B'>
F2: <function _create_type at 0x10681c158>
T1: <class 'type'>
F2: <function _load_type at 0x10681c0d0>
T1: <class 'object'>
D2: <dict object at 0x10431b808>
B1: <built-in function _reconstruct>
F2: <function _get_attr at 0x10681ca60>
M1: <module 'numpy.core.multiarray' from '/Users/itdxer/.pyenv/versions/neuralpy/lib/python3.4/site-packages/numpy/core/multiarray.so'>
F2: <function _import_module at 0x10681cb70>
D2: <dict object at 0x10431b488>
B1: <built-in function is_busday>
B1: <built-in function datetime_as_string>
B1: <built-in function _get_ndarray_c_version>
T4: <class 'PyCapsule'>
Traceback (most recent call last):
  File "/Users/itdxer/.pyenv/versions/3.4.3/lib/python3.4/pickle.py", line 272, in _getattribute
    obj = getattr(obj, subpath)
AttributeError: 'module' object has no attribute 'PyCapsule'
```

**Comment by matsjoyce** [comment]
@itdxer Could you try the patch at #92, because this looks similar, as the library is in the virtual env. What's the value of `sys.base_prefix` and `sys.real_prefix`?

**Comment by itdxer** [comment]
I install last version from master master (0.2.3.dev0), but it didn't help. I modify code and change value from `real_prefix` to `prefix` and this update helps me. `prefix` attribute from module `sys` contains path to my virtualenv and `sys.real_prefix` raise AttributeError exception when I try check it from terminal.
I use [pyenv](https://github.com/yyuu/pyenv) and virtualenv inside of it.

**Comment by mmckerns** [comment]
@matsjoyce: I hate to say this, but what about this being another case of #86 -- assuming the `D2: <dict object at 0x10431b488>` is a `weakref` or something similar that points to `globals`?

**Comment by matsjoyce** [comment]
I think the key difference is that one of his lines is:
```
M1: <module 'numpy.core.multiarray' from '/Users/itdxer/.pyenv/versions/env/lib/python3.4/site-packages/numpy/core/multiarray.so'>
```
While yours is:
```
M2: <module 'numpy.core.multiarray' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/numpy/core/multiarray.so'>
```
I think its similar to #91, but with a different type of virtual env. TO fix this I should probably extend #92 to all the `sys.*prefix`s because @itdxer said:

> I modify code and change value from real_prefix to prefix and this update helps me.

I'll have a go later.

**Comment by mmckerns** [comment]
Ah… I missed that.  I expect you're right.
